components/director/pkg/operation/middleware.go

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2020 The Compass Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package operation
+
+import (
+	"context"
+	"fmt"
+
+	gqlgen "github.com/99designs/gqlgen/graphql"
+	"github.com/kyma-incubator/compass/components/director/pkg/graphql"
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+	"github.com/tidwall/sjson"
+	"github.com/vektah/gqlparser/ast"
+)
+
+const LocationsParam = "locations"
+
+type middleware struct {
+	directorURL string
+}
+
+// NewMiddleware creates a new handler struct responsible for enriching the response of Async mutations with Operation URL location information
+func NewMiddleware(directorURL string) *middleware {
+	return &middleware{
+		directorURL: directorURL,
+	}
+}
+
+// ExtensionHandler enriches Async mutation responses with Operation URL location information and also empties the data property of the graphql response for such requests
+func (m *middleware) ExtensionHandler(ctx context.Context, next func(ctx context.Context) []byte) []byte {
```

Comment:
can we not split into two but instead have all logic for operation pre and post processing in one place (in this case the directive)? I suspect it would be even simpler code there, no need for sjson deletion, just returning empty bytes

Reply:
The initial idea was that if there are multiple gql mutations in a single request, this middleware would be called only once while the directives + resolvers are called for every mutation - this would allow us to initialize the the Operations slize in the context in a single place and also register the extensions in one shot. Maybe we can think of a way to move this in the directive, not sure if we'll be able to build on the registered extensions, but we'll try it out